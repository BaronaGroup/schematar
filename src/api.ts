import fs from 'fs'
import path from 'path'

import mkdirp from 'mkdirp'

import ComplexType from './complex'
import { generateHash } from './hash-schema'
import jsonSchemaOutput, { JSONSchemaOptions } from './json-schema-output'
import { karhu } from './karhu'
import mongooseOutput from './mongoose-output'
import nowToken from './now'
import { ObjectId as ObjectIdType } from './object-id'
import { Schema as SchemaType } from './schema'
import tsOutput, { TSOptions } from './ts-output'

export const ObjectId = ObjectIdType
export type Schema = SchemaType
export const Complex = ComplexType
export const now = nowToken
const log = karhu('schematar-api')

export function createMongooseSchema(schema: Schema, context = 'mongoose') {
  return mongooseOutput(schema, context)
}

export function createTypescriptInterfaceDefinition(
  exportName: string,
  schema: Schema | ComplexType,
  context: string = 'typescript',
  options: TSOptions = {}
) {
  return tsOutput(exportName, schema, context, options)
}

export function createJSONSchema(schema: Schema, context: string = 'jsonschema', options: JSONSchemaOptions = {}) {
  return jsonSchemaOutput(schema, context, options)
}

export async function createTypescriptInterfaceFiles(
  sourceFileGlobOrFileArray: string | string[],
  outputPath: string,
  logCreations = false
) {
  const { default: glob } = await import('glob')
  await new Promise((resolve, reject) => mkdirp(outputPath, (err) => (!err ? resolve() : reject(err))))
  const files = [].concat(
    ...(Array.isArray(sourceFileGlobOrFileArray)
      ? sourceFileGlobOrFileArray.map((f) => glob.sync(f))
      : [glob.sync(sourceFileGlobOrFileArray)])
  )
  for (const file of files) {
    await createTSInterfaceFile(file, outputPath, logCreations)
  }
}

export interface TypescriptSchemaDefinition {
  name: string
  context?: string
  omitExtraExports?: boolean
  schema?: Schema | ComplexType
  exportHash?: string
  doNotImportObjectId?: boolean
}

interface SchemaFile {
  name?: string
  default?: Schema
  schema?: Schema
  typescriptSchemas?: TypescriptSchemaDefinition[]
}

export async function createTSInterfaceFile(
  filename: string,
  outputPath: string,
  logCreations: boolean = true,
  formatOutputFilename = defaultOutputFilenameFormatter
) {
  const schemaFile = require(path.resolve(filename)) as SchemaFile
  const topLevelSchema = schemaFile.schema || schemaFile.default
  const topLevelName = schemaFile.name || pickNameFromFilename(filename)
  const exported: string[] = []
  if (schemaFile.typescriptSchemas) {
    const typescriptSchemas = schemaFile.typescriptSchemas
    for (const schemaDef of typescriptSchemas) {
      const schema = schemaDef.schema || topLevelSchema,
        name = schemaDef.name || topLevelName

      if (!schema) throw new Error('No schema found for ' + name)
      exported.push(
        createTypescriptInterfaceDefinition(name, schema, schemaDef.context || 'typescript', {
          ...schemaDef,
          doNotImportObjectId: schemaDef.doNotImportObjectId || typescriptSchemas.indexOf(schemaDef) > 0,
        })
      )
    }
  } else if (topLevelSchema) {
    exported.push(createTypescriptInterfaceDefinition(topLevelName, topLevelSchema))
  }
  const outputFilename = path.join(outputPath, formatOutputFilename(filename))

  if (logCreations) {
    log.info('Storing typescript interfaces in', outputFilename)
  }

  if (exported.length) {
    await new Promise((resolve, reject) =>
      fs.writeFile(outputFilename, exported.join('\n'), 'utf8', (err) => (!err ? resolve() : reject(err)))
    )
  }
}

function defaultOutputFilenameFormatter(inputFilename: string) {
  const ext = path.extname(inputFilename),
    basename = path.basename(inputFilename, ext)

  return basename + '.schematar.ts'
}

function pickNameFromFilename(filename: string) {
  const fn = path.basename(filename, path.extname(filename))
  return (fn[0].toUpperCase() + fn.substring(1))
    .replace(/[\-_][a-z]/g, (match) => match[1].toUpperCase())
    .replace(/[^a-zA-Z0-9]/g, '')
}

export function hashSchema(schema: Schema | ComplexType) {
  return generateHash(schema)
}
