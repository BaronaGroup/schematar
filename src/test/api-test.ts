import {createTypescriptInterfaceFiles} from '../api'
import fs from 'fs'
import mkdirp = require('mkdirp')

const isOutputFile = /\.schematar\.ts$/

describe('api-test', function() {
  const outputDir = __dirname + '/../../test-output/api-test'
  beforeAll(async () => {
    await prepareOutputDirectory(outputDir)
    await createTypescriptInterfaceFiles(__dirname + '/test-schemas/*-schema.ts', outputDir)
  })

  describe('createTypescriptInterfaceFiles', function() {
    describe('implicit', function() {
      it('simple case', testOutputSnapshot('simple'))
      it('simple case with default export', testOutputSnapshot('default'))
      it('simple case with explicit name', testOutputSnapshot('named'))
    })

    describe('explicit', function() {
      it('no outputs', function() {
        expect(fs.existsSync(getOutputFilename('no-ts'))).toBe(false)
      })

      it('explicit name', testOutputSnapshot('exp-name'))

      it('explicit schema', testOutputSnapshot('exp-schema'))

      it('explicit context', testOutputSnapshot('exp-context'))

      it('omit extra exports', testOutputSnapshot('no-extra-exports'))

      it('export schema hash', testOutputSnapshot('export-hash'))

      it('multiple schemas', testOutputSnapshot('multiple-schemas'))

      it('supports using complex in place of the schema', testOutputSnapshot('complex-as-schema'))
    })
  })

  function getOutputFilename(baseName: string) {
    return `${outputDir}/${baseName}-schema.schematar.ts`
  }

  function testOutputSnapshot(baseName: string) {
    return async function() {

      const data = await fs.promises.readFile(getOutputFilename(baseName), 'utf8')
      expect(data).toMatchSnapshot()
    }
  }
})

async function prepareOutputDirectory(dir: string) {
  if (!fs.existsSync(dir)) {
    mkdirp.sync(dir)
  }
  for (const file of fs.readdirSync(dir)) {
    const ffn = `${dir}/${file}`
    if (isOutputFile.test(ffn)) {
      fs.unlinkSync(ffn)
    }
  }
}
