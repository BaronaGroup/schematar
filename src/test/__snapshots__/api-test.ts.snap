// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`api-test createTypescriptInterfaceFiles explicit explicit context 1`] = `
"// Generated file, do not edit!

// tslint:disable array-type
// @ts-ignore -- ignore possibly unused type parameters
export interface ExpContextSchemaBase<IDType, DateType>
{
  field1: string
  field2?: string
  field3: string
  field4: string
}
import {ObjectId} from 'mongodb'
export type ExpContextSchemaMongoose = ExpContextSchemaBase<ObjectId, Date>
export type ExpContextSchemaJSON = ExpContextSchemaBase<string, string>
export type ExpContextSchemaFluid = ExpContextSchemaBase<string | ObjectId, string | Date>"
`;

exports[`api-test createTypescriptInterfaceFiles explicit explicit name 1`] = `
"// Generated file, do not edit!

// tslint:disable array-type
// @ts-ignore -- ignore possibly unused type parameters
export interface CustomNameBase<IDType, DateType>
{
  field: string
}
import {ObjectId} from 'mongodb'
export type CustomNameMongoose = CustomNameBase<ObjectId, Date>
export type CustomNameJSON = CustomNameBase<string, string>
export type CustomNameFluid = CustomNameBase<string | ObjectId, string | Date>"
`;

exports[`api-test createTypescriptInterfaceFiles explicit explicit schema 1`] = `
"// Generated file, do not edit!

// tslint:disable array-type
// @ts-ignore -- ignore possibly unused type parameters
export interface ExpSchemaSchemaBase<IDType, DateType>
{
  altField: number
}
import {ObjectId} from 'mongodb'
export type ExpSchemaSchemaMongoose = ExpSchemaSchemaBase<ObjectId, Date>
export type ExpSchemaSchemaJSON = ExpSchemaSchemaBase<string, string>
export type ExpSchemaSchemaFluid = ExpSchemaSchemaBase<string | ObjectId, string | Date>"
`;

exports[`api-test createTypescriptInterfaceFiles explicit export schema hash 1`] = `
"// Generated file, do not edit!

// tslint:disable array-type
// @ts-ignore -- ignore possibly unused type parameters
export interface ExportHashSchemaBase<IDType, DateType>
{
  field: string
}
import {ObjectId} from 'mongodb'
export type ExportHashSchemaMongoose = ExportHashSchemaBase<ObjectId, Date>
export type ExportHashSchemaJSON = ExportHashSchemaBase<string, string>
export type ExportHashSchemaFluid = ExportHashSchemaBase<string | ObjectId, string | Date>
export const mySchemaHash = '5d3861b32891d91c94cafb81aa1f821ae442e516d406e1e3bea47f6cb312b11f'"
`;

exports[`api-test createTypescriptInterfaceFiles explicit multiple schemas 1`] = `
"// Generated file, do not edit!

// tslint:disable array-type
// @ts-ignore -- ignore possibly unused type parameters
export interface MultipleSchemasSchemaBase<IDType, DateType>
{
  field: string
}
import {ObjectId} from 'mongodb'
export type MultipleSchemasSchemaMongoose = MultipleSchemasSchemaBase<ObjectId, Date>
export type MultipleSchemasSchemaJSON = MultipleSchemasSchemaBase<string, string>
export type MultipleSchemasSchemaFluid = MultipleSchemasSchemaBase<string | ObjectId, string | Date>
// Generated file, do not edit!

// tslint:disable array-type
// @ts-ignore -- ignore possibly unused type parameters
export interface AlternativeBase<IDType, DateType>
{
  another: number
}
export type AlternativeMongoose = AlternativeBase<ObjectId, Date>
export type AlternativeJSON = AlternativeBase<string, string>
export type AlternativeFluid = AlternativeBase<string | ObjectId, string | Date>
export const anotherHash = '5769d7ffccddeec74191da4ca235bf7efb58994cb7635815af5d02ae8008c610'
// Generated file, do not edit!

// tslint:disable array-type
// @ts-ignore -- ignore possibly unused type parameters
export interface FinalBase<IDType, DateType>
{
  field: string
  finalField: string
}
export type FinalMongoose = FinalBase<ObjectId, Date>
export type FinalJSON = FinalBase<string, string>
export type FinalFluid = FinalBase<string | ObjectId, string | Date>"
`;

exports[`api-test createTypescriptInterfaceFiles explicit omit extra exports 1`] = `
"// Generated file, do not edit!

// tslint:disable array-type
// @ts-ignore -- ignore possibly unused type parameters
export interface NoExtraExportsSchemaBase<IDType, DateType>
{
  field: string
}"
`;

exports[`api-test createTypescriptInterfaceFiles explicit supports using complex in place of the schema 1`] = `
"// Generated file, do not edit!

// tslint:disable array-type
// @ts-ignore -- ignore possibly unused type parameters
export interface CustomNameBase<IDType, DateType>
{
    field: string
    field2: number
  }
import {ObjectId} from 'mongodb'
export type CustomNameMongoose = CustomNameBase<ObjectId, Date>
export type CustomNameJSON = CustomNameBase<string, string>
export type CustomNameFluid = CustomNameBase<string | ObjectId, string | Date>"
`;

exports[`api-test createTypescriptInterfaceFiles implicit simple case 1`] = `
"// Generated file, do not edit!

// tslint:disable array-type
// @ts-ignore -- ignore possibly unused type parameters
export interface SimpleSchemaBase<IDType, DateType>
{
  field: string
}
import {ObjectId} from 'mongodb'
export type SimpleSchemaMongoose = SimpleSchemaBase<ObjectId, Date>
export type SimpleSchemaJSON = SimpleSchemaBase<string, string>
export type SimpleSchemaFluid = SimpleSchemaBase<string | ObjectId, string | Date>"
`;

exports[`api-test createTypescriptInterfaceFiles implicit simple case with default export 1`] = `
"// Generated file, do not edit!

// tslint:disable array-type
// @ts-ignore -- ignore possibly unused type parameters
export interface DefaultSchemaBase<IDType, DateType>
{
  field: string
}
import {ObjectId} from 'mongodb'
export type DefaultSchemaMongoose = DefaultSchemaBase<ObjectId, Date>
export type DefaultSchemaJSON = DefaultSchemaBase<string, string>
export type DefaultSchemaFluid = DefaultSchemaBase<string | ObjectId, string | Date>"
`;

exports[`api-test createTypescriptInterfaceFiles implicit simple case with explicit name 1`] = `
"// Generated file, do not edit!

// tslint:disable array-type
// @ts-ignore -- ignore possibly unused type parameters
export interface BananaBase<IDType, DateType>
{
  field: string
}
import {ObjectId} from 'mongodb'
export type BananaMongoose = BananaBase<ObjectId, Date>
export type BananaJSON = BananaBase<string, string>
export type BananaFluid = BananaBase<string | ObjectId, string | Date>"
`;
